corehq/apps/sso/backends.py

```diff
@@ -0,0 +1,57 @@
+from django.contrib.auth.backends import ModelBackend
+from django.contrib.auth.models import User
+
+from corehq.apps.sso.models import IdentityProvider, AuthenticatedEmailDomain
+
+
+class SsoBackend(ModelBackend):
+    """
+    Authenticates against an IdentityProvider and SAML2 session data.
+    """
+
+    def authenticate(self, request, username=None, idp_slug=None, **kwargs):
```

Comment:
won't removing the `kwarg` defaults result in a 500 when django loops through all the backends trying to match one with an `authenticate` request?

Reply:
Oh, I didn't know this will be called by code we did not write. Fine to keep the defaults if the interface it must provide includes being called without those args.

Reply:
Another question related to https://github.com/dimagi/commcare-hq/pull/29215#discussion_r581345588, will Django ever call this with `username` and `idp_slug`? If not, then why are we exposing it to Django as an authentication backend (since authentication will always fail when Django calls it)?

Reply:
that's a good question. I think I was just trying to follow the django pattern for authentication. If I should pull this out entirely of being a django-style backend, then I should just do that I suppose...thoughts? Then there would be absolutely zero chance of this being called outside of saml's `acs` view

Reply:
Yes, that seems like the right thing to do.